Adding a config setting for allowing some attributes to filter targets when in hints mode number

mariusor opened the pull request:
This would close issue #4634.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/4639)
<!-- Reviewable:end -->


[edit by @jgkamat to link issues: Closes #4634]

jgkamat commented on qutebrowser/browser/hints.py:
```diff
@@ -570,23 +570,27 @@ def _check_args(self, target, *args):
+        elemstr = str(elem).casefold()
+        try:
+            title = elem.title().casefold()
+            elemstr += " " + title
```
I don't think this is the right place to add this to the elemstr - it would make more sense for people to pass this function the extended list. Even better, a list of words to match on.

I'm not sure where best to do that though. Maybe a seperate function to generate an elemstr from an element would be useful.

mariusor replied:
What do you mean by "a list of words to match on" ?

If you mwan which attributes of the element to match against, then the list is pretty limited, we have `title` and `alt`.

Another solution to the issue, would be to modify the method that converts an element to string to include these attributes if they exist. 

Let me know which one would be preferred.

jgkamat replied:
I don't mean a list of elements, but a list of the attributes. Eg: if an element's title is "my-title" and alt is "my-alt", I would expect ["my-title", "my-alt"] to be passed in.

I don't really care about a list vs a string though, I just don't think `elemstr` should be touched in this function.

mariusor replied:
Please take a look at the updated code, this diff doesn't correspond to the code any more. The current version should match what you mentioned.

mariusor commented on qutebrowser/browser/webkit/webkitelem.py:
```diff
@@ -111,6 +111,10 @@ def tag_name(self) -> str:
 
+    def title(self) -> str:
+        """Get the title for the current element."""
+        return self._js_dict['attributes']['title']
```
Makes perfect sense. I will do that.

toofar commented:
I still don't get the use case. Aren't element titles not visible unless you hover them? And at that point you might as well have already selected them.
Also selecting links based on non-visible text may well break the workflow of people who don't want this behaviour because there might type out the text they see expecting that element to be selected but if there is something with a matching title then it wont.

mariusor commented:
> I still don't get the use case.

For links that don't have text as children, but an image (for example icons that perform an action: edit, delete, home, etc), they usually have titles, which specify their use and which most of the time are self evident, or easily remembered.

> Also selecting links based on non-visible text may well break the workflow of people who don't want this behaviour because there might type out the text they see expecting that element to be selected but if there is something with a matching title then it wont.

That is simply not true. The links which are currently highlighted (by their text content) would still be highlighted. In the case of this PR there would be more links highlighted, and this would maybe break the use case of someone expecting a specific filter text to automatically follow a specific link, which in this case, if the filter text matched the title of other links, would show them all and wait for a number decision.

For this case, I would be OK to hide this functionality behind a configuration toggle.

jgkamat commented on qutebrowser/config/configdata.yml:
```diff
@@ -1091,6 +1091,14 @@ hints.mode:
       - word: Use hints words based on the html elements and the extra words.
   desc: Mode to use for hints.
 
+hints.extended_filters:
```
This shouldn't be a boolean, but instead a list of strings (enum). That way, more types can be added later on.

(I still don't think elemstr should be modified in `_filter_matches`)

mariusor replied:
Replaced with a list.

jgkamat commented on qutebrowser/browser/hints.py:
```diff
@@ -179,6 +180,7 @@ class HintContext:
     group = attr.ib(None)
     hint_mode = attr.ib(None)
     first = attr.ib(False)
+    extended_filter_targets = False
```
Why save this instead of just pulling it from config directly?

mariusor replied:
Removed.

mariusor commented:
I don't know what "." means. :)

jgkamat commented on qutebrowser/browser/hints.py:
```diff
@@ -570,23 +570,31 @@ def _check_args(self, target, *args):
+        matchagainst = elemstr
+        for _attr in config.val.hints.extended_filter_attrs:
+            try:
+                matchagainst += ' ' + elem[_attr].casefold()
```
This is called quite a bit (number of hints * number of filter attributes), so I would really like to avoid the exception here. It should be trivial to implement a `get` method in things that implement AbstractWebElement, could you do that?

jgkamat commented on qutebrowser/browser/hints.py:
```diff
@@ -570,23 +570,31 @@ def _check_args(self, target, *args):
-    def _filter_matches_exactly(self, filterstr, elemstr):
-        """Return True if `filterstr` exactly matches `elemstr`."""
+        matchagainst = elemstr
+        for _attr in config.val.hints.extended_filter_attrs:
```
Can you use `config.cache['hints.extended_filter_attrs']` here instead? Currently it will do a loop over every char in every string in the array (which is called on every hint) which can be costly

toofar commented:
> would maybe break the use case of someone expecting a specific filter text to automatically follow a specific link

Yeah that what I was thinking about, muscle memory power users stuff.

> For links that don't have text as children

Okay, would falling back to the title if `str(elem)` is empty be sufficient for you?

Note that I am not asking you to make changes based on my questions, I don't even use number hinting. I'm just asking you to consider other workflows that might be affected. Like for instance hints don't just match links but all sorts of elements.

mariusor commented:
@jgkamat 
> I still don't think that the attributes should be built inside `_filter_matches`

I added the method `matches_hint_filter` on `WebEngineElement` and `WebEngineElement`. I pushed the matching logic there, but I'm not sure that's what you wanted. :( Please give me more guidance if not. 

> Can you add some tests for this? There's some examples in hints.feature in the 'Renumbering hints when filtering' test. 

Done, but still working on making them pass. :))

> Not sure what we should do by default here, I'm no good at setting defaults, but I wanted to bring it up.
> Maybe we could make text a valid option for the filter attributes as well, making it possible to turn off filtering completely.

I added the `text` value as an option and leaving it as default for the `hints.extended_filter_attrs` config setting. This means that the default behaviour of `hints.mode: number` will match perfectly the current one. This should address @toofar 's questions about existing workflows also.

mariusor commented:
@jgkamat honestly I don't know if the tests that are failing are doing so because of my test or not. Are there some docs to run the behat tests locally ?

jgkamat commented:
Marius Orcsik writes:

> @jgkamat honestly I don't know if the tests that are failing are doing so because of my test or not. Are there some docs to run the behat tests locally ?

Yup, the contributing docs have instructions. I don't reccomend running all
the tests, as they are quite slow.

https://github.com/qutebrowser/qutebrowser/blob/master/doc/contributing.asciidoc#running-specific-tests

Looking at travis, I think
'tests/end2end/features/test_hints_bdd.py::test_using_hintsmode_number_with_title_attribute_filtering'
is failing on webkit, and there's a bunch of style issues (but resolving them
is usually a bunch of busywork, so I usually wait for that until the end to
minimize the time wasted).

> I added the method `matches_hint_filter` on `WebEngineElement` and `WebEngineElement`. I pushed the matching logic there, but I'm not sure that's what you wanted. :( Please give me more guidance if not.

Yup, this looks like what I was thinking of, nice :]. I'll try to get another
real review in soon.

jgkamat commented on qutebrowser/browser/webkit/webkitelem.py:
```diff
@@ -111,6 +111,25 @@ def tag_name(self) -> str:
 
+    def attr(self, name: str) -> str:
+        """Get 'name' attribute if exists"""
+        return self._js_dict.get('attributes', {}).get(name, '')
```
I don't think webkit has an element for `_js_dict`. You should be able to copy `__getitem__` with a default argument here though

jgkamat commented on qutebrowser/browser/webkit/webkitelem.py:
```diff
@@ -111,6 +111,25 @@ def tag_name(self) -> str:
+        """Get 'name' attribute if exists"""
+        return self._js_dict.get('attributes', {}).get(name, '')
+
+    def matches_hint_filter(self, filterstr: str, attrs=['text']) -> str:
```
I think the implementation between webkit/webengine is the same here, so I think the concrete def can go in AbstractWebElement so we only need one (just make sure to use non-specific methods, which I don't think you are)

jgkamat commented on qutebrowser/browser/webkit/webkitelem.py:
```diff
@@ -111,6 +111,25 @@ def tag_name(self) -> str:
+            else:
+                matchstr = self.attr(_attr).casefold()
+
+            if all(word in matchstr for word in filterstr.split()):
```
Can you lift this split outside the loop, so it's not calculated for every attribute? Even better to split it before passing it in, although the debug hint might change a little (so I'm fine with leaving that off for now).

mariusor commented:
@jgkamat thank you for all the feedback. 

I had some struggles in getting the linters pass, but I think we're good now. Can you have a look at the test that failed in the last build and tell me if it's probable that it happened because of my changes, or is just a timeout due to low resources on travis?

mschilli87 commented on qutebrowser/config/configdata.yml:
```diff
@@ -1091,6 +1090,21 @@ hints.mode:
+        - title: "the 'title' attribute"
+        - alt: "the 'alt' attribute"
+  desc: >-
+    Specify which HTML element attributes to use additionally when filtering links in `hints.mode` number.
```
```suggestion
    Which HTML element attributes to use additionally when filtering links in
    `hints.mode` number.
```

jgkamat commented on qutebrowser/browser/webelem.py:
```diff
@@ -114,6 +114,31 @@ def tag_name(self) -> str:
+        return False
+
+    def attr(self, name: str) -> str:
+        """Get 'name' attribute if exists."""
```
Nit: Can you mention that '' is the default value? Strictly, you don't need docs on the override methods (since they're just implementing this one), but please update them too if you keep them.

toofar commented on qutebrowser/config/configdata.yml:
```diff
@@ -766,7 +766,6 @@ content.webrtc_ip_handling_policy:
       - disable-non-proxied-udp: WebRTC should only use TCP to contact peers or
             servers unless the proxy server supports UDP. This doesn't expose
             any local addresses either.
-  default: all-interfaces
```
This shouldn't have been removed.

mariusor replied:
It was marked as a duplicate key in my editor. But yeah, it's not related to the changes.

mariusor commented:
@toofar I'm not sure what you're asking in your last comment. Can you clarify a bit? I modified the documentation of `attr` but I'm not sure this was all you wanted. :D

toofar commented:
I mean I don't think you need to add those new `attr` methods since `__getitem__` already does almost the same thing.
```python
diff --git i/qutebrowser/browser/webelem.py w/qutebrowser/browser/webelem.py
index 7268247ff3ad..50f25bd00dc9 100644
--- i/qutebrowser/browser/webelem.py
+++ w/qutebrowser/browser/webelem.py
@@ -127,7 +127,10 @@ class AbstractWebElement(collections.abc.MutableMapping):
             if _attr == 'text':
                 matchstr = str(self).casefold()
             else:
-                matchstr = self.attr(_attr).casefold()
+                try:
+                    matchstr = self[_attr].casefold()
+                except KeyError:
+                    pass
 
             if all(word in matchstr for word in _words):
                 log.webelem.debug(

```
Except it raises a KeyError instead of returning `''` which means you don't have to bother trying to match `_words` against a bunch of empty strings each time.

toofar commented:
Okay, ignore my previous comment. I wasn't following the other reviews closely enough. Instead can you just add a `if not matchstr: continue` check before if actually does the matching because that case should never match anything anyway because the method only gets called after text is entered.

mariusor commented:
> Instead can you just add a `if not matchstr: continue` check before it actually does the matching? 

Done. Thanx.

mariusor commented:
Any more steps I need to do, in order to get this merged? :)

The-Compiler commented:
Not at the moment - I'd like to take a look at well, but that might take a while, as there is a quite big backlog of PRs.

mariusor commented:
Hey @The-Compiler. Any news? :)

The-Compiler commented:
Not yet, there's still quite a backlog as you can see :wink: I had to ignore pull requests for quite a long time due to an university project related to qutebrowser, see the [notice in the contribution docs](https://github.com/qutebrowser/qutebrowser/commit/e390d5a1a2fda0a0e6ae26d6521311b12cf05430#diff-1b6beb8ff5b36bb30a4a97474054f858) (which was since reworded)

mariusor commented:
OK, I see. Thanx.

mariusor commented:
I'm not sure if there's still interest to merge this PR, but I updated it so it doesn't conflict with master any more.

The-Compiler commented:
There is! I'm currently setting a few things up, then I should finally have time to look at a couple of PRs every week and get the backlog down - see the recent [roadmap mail](https://s.cmpl.cc/qute201910) for details.

mariusor commented:
Any news @The-Compiler ? We could have had a baby since I submitted this. :)

mschilli87 commented on qutebrowser/config/configdata.yml:
```diff
@@ -1100,6 +1100,22 @@ hints.mode:
+      valid_values:
+        - text: "the actual node text"
+        - title: "the 'title' attribute"
+        - alt: "the 'alt' attribute"
```
Could you turn these into sentences? Just have a look at other `valid_values` definitions in this file. This is just for consistency and make hings more self-explanatory. Also, if the text doesn't include any `'`, there is no need to wrap it in `"`s.

mariusor replied:
Is it OK like this?

mariusor commented:
I'm closing this because @The-Compiler considers I'm being passive aggressive about mentioning an almost year old pull request.

I would suggest other people take responsibility for merging PRs because this is not exactly a model of good cooperation.

Cheers.

The-Compiler commented:
For reference:

https://www.reddit.com/r/qutebrowser/comments/dsz76a/are_there_any_qutebrowser_users_who_live_in_berlin/f6t2vaf

https://www.reddit.com/r/qutebrowser/comments/ey3yjq/qutebrowser_v1100_released/fgf1q14

Call it a joke if you will, but yes, I do consider "merge my PR" comments on totally unrelated topics passive-aggressive. I also don't consider them cooperative in any way.
